crates/iceberg-catalog/src/service/storage/gcs/sts.rs

```diff
@@ -0,0 +1,226 @@
+        .send()
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to send downscoping request: {:?}", e);
+            TableConfigError::FailedDependency("Failed to send downscoping request".to_string())
+        })?
+        .json::<STSResponse>()
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to parse downscoping response: {:?}", e);
+            TableConfigError::StsError("Downscoping failed".to_string(), Some(Box::new(e)))
+        })
+}
+
+#[derive(Deserialize, veil::Redact)]
+pub(crate) struct STSResponse {
+    #[redact(partial)]
+    pub(crate) access_token: String,
+    pub(crate) expires_in: Option<usize>,
+    token_type: String,
+}
+
+#[derive(Serialize)]
+struct STSRequest {
+    // urn:ietf:params:oauth:grant-type:token-exchange
+    pub grant_type: String,
+    /// The full resource name of the identity provider; for example:
+    /// //iam.googleapis.com/projects/<project-number>/locations/global/workloadIdentityPools/<pool-id>/providers/<provider-id>
+    /// for workload identity pool providers, or
+    /// //iam.googleapis.com/locations/global/workforcePools/<pool-id>/providers/<provider-id> for
+    /// workforce pool providers. Required when exchanging an external credential for a Google
+    /// access token.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub audience: Option<String>,
+    /// The OAuth 2.0 scopes to include on the resulting access token, formatted as a list of space-
+    /// delimited, case-sensitive strings. Required when exchanging an external credential for a
+    /// Google access token.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub scope: Option<String>,
+    // urn:ietf:params:oauth:token-type:access_token
+    pub requested_token_type: String,
+    pub subject_token: String,
+    pub subject_token_type: String,
+    // serialized json string
+    pub options: String,
+}
+
+impl STSRequest {
+    fn from_token_and_options(token: &str, options: &Options) -> Result<Self, TableConfigError> {
+        let op = serde_json::to_string(options).map_err(|e| {
+            TableConfigError::StsError("Failed to serialize options".to_string(), Some(Box::new(e)))
+        })?;
+        Ok(Self {
+            grant_type: "urn:ietf:params:oauth:grant-type:token-exchange".to_string(),
+            audience: None,
+            scope: None,
+            requested_token_type: "urn:ietf:params:oauth:token-type:access_token".to_string(),
+            subject_token: token.to_string(),
+            subject_token_type: "urn:ietf:params:oauth:token-type:access_token".to_string(),
+            // string
+            //
+            // A JSON-format Credential Access Boundary, encoded with percent encoding.
+            // idk?
+            options: percent_encoding::utf8_percent_encode(&op, percent_encoding::NON_ALPHANUMERIC)
+                .to_string(),
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct Options {
```

Comment:
Shouldn't we also add an audience to scope it to the storage service?
https://cloud.google.com/iam/docs/reference/sts/rest/v1beta/Options

Reply:
Not sure what's going on there but the endpoint rejected that:

```
downscoping response: "{\"error\":\"invalid_request\",\"error_description\":\"Unsupported argument: `options.audiences`\"}"
```

Reply:
even google's own python implementation is not exposing that parameter, https://github.com/googleapis/google-auth-library-python/blob/484c8db151690a4ae7b6b0ae38db0a8ede88df69/google/auth/downscoped.py#L170

Reply:
Well, then let's forget it. Strange